Initialize LSP progress token before using it and remove progress for sync plugins

syphar opened the pull request:
This now initializes a progress token before using it. 

I'm don't have much experience with json-rpc & LSP implementation details, the adapted tests & fixtures stem from my little understanding. 

I'm happy to add any needed adaptions. 

Fixes #325

syphar committed:
fixes #325: initialize LSP progress token before using it

ccordoba12 commented on test/fixtures.py:
```diff
@@ -5,7 +5,9 @@
 from io import StringIO
 from unittest.mock import MagicMock
 import pytest
```
```suggestion

import pytest
```

We accustom to separate standard library, third-party and local imports with blank lines.

syphar replied:
fixed. 

In general, are you open to introduce `isort` into the linting config for this? 

IMO this eases contributions / reviews with many contributors. 

And while we're on it, what about black? 

( in any case, both not in this PR)

ccordoba12 replied:
> In general, are you open to introduce isort into the linting config for this?

Yes, I am. Thanks for suggesting it.

> And while we're on it, what about black?

Sure, sounds good to me too.

> ( in any case, both not in this PR)

Agreed. I don't have time to do it tough, but I'll be happy to review and merge such contribution.

ccordoba12 commented on test/fixtures.py:
```diff
@@ -63,14 +65,55 @@ def pylsp_w_workspace_folders(tmpdir):
     return (ls, workspace_folders)
 
 
+class FakeEditorMethodsMixin:
```
Please move all the new supporting classes you added here to be before the fixture declarations.

syphar replied:
done

ccordoba12 commented on pylsp/_utils.py:
```diff
@@ -14,6 +14,7 @@
 import jedi
 
 JEDI_VERSION = jedi.__version__
+CALL_TIMEOUT = 10
```
Perhaps this constant should have a more specific name? I mean, `CALL_TIMEOUT` sounds too generic to me. What about `PROGRESS_CALL_TIMEOUT`?

syphar replied:
I removed the need for it because I removed the usage of this timeout for progress. 

It's back having the same name as before in `tests.test_language_server`

ccordoba12 reviewed (commented):
Thanks @syphar for your prompt response to address issue #325! I left some small suggestions for you, otherwise looks good to me.

ccordoba12 commented:
@rchl, could you take a look at this one? Thanks!

rchl commented on pylsp/workspace.py:
```diff
@@ -152,6 +153,9 @@ def _progress_begin(
     ) -> str:
         token = str(uuid.uuid4())
+
+        self._endpoint.request(self.M_INITIALIZE_PROGRESS, {'token': token}).result(_utils.CALL_TIMEOUT)
```
To quote the spec:

> To keep the protocol backwards compatible servers are only allowed to use `window/workDoneProgress/create` request if the client signals corresponding support using the client capability `window.workDoneProgress`
>
> https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#serverInitiatedProgress

And also it's a request so the server needs to wait for the result and is then only allowed to use that token if the result is successful.

syphar replied:
that's a valid point. I also checked rust-analyzer source, they also do it. 

I added checking for the client capability.

rchl commented on pylsp/workspace.py:
```diff
@@ -152,6 +153,9 @@ def _progress_begin(
     ) -> str:
         token = str(uuid.uuid4())
+
+        self._endpoint.request(self.M_INITIALIZE_PROGRESS, {'token': token}).result(_utils.CALL_TIMEOUT)
```
I'm not sure why would you use a timeout for this specific request. Any compliant client should either respond with an OK or error response.

Though if you are not checking for client capability then all bets are off but that would be a server issue then.

syphar replied:
valid point, I removed the timeout.

syphar committed:
respect window.workDoneProgress capability to activate progress reporting

syphar committed:
don't timeout for progress-reporting RPC requests to editor

syphar committed:
move test-helper classes to top of fixture file

syphar committed:
fix import grouping / format

syphar committed:
fix pylint errors

rchl commented on pylsp/workspace.py:
```diff
@@ -152,6 +167,9 @@ def _progress_begin(
     ) -> str:
         token = str(uuid.uuid4())
+
+        self._endpoint.request(self.M_INITIALIZE_PROGRESS, {'token': token}).result()
```
This is not being awaited, is it?

The server must wait for the response and in case of an error not send progress reports.

syphar replied:
that's what `.result` does.

syphar replied:
https://docs.python.org/3.11/library/concurrent.futures.html#concurrent.futures.Future.result

syphar replied:
and https://github.com/python-lsp/python-lsp-jsonrpc/blob/c73fbdba2eeb99b7b145dcda76e62250552feda4/pylsp_jsonrpc/endpoint.py#L60-L87

rchl replied:
I see. It's hard to know that since many of the types in this file are missing/Unknown.

That said, I see a type issue that this change introduces.

In `python_lsp.py` the `self._endpoint` should be typed as `Endpoint` and it is on `develop` branch:

![Screenshot 2023-01-02 at 20 54 10](https://user-images.githubusercontent.com/153197/210273704-b35a6563-a73f-4211-9dca-54b94ae8d4bf.png)

But with those changes it's `Unknown | Endpoint`:

![Screenshot 2023-01-02 at 20 56 46](https://user-images.githubusercontent.com/153197/210273736-de7540cf-3726-4472-803e-1095c0338805.png)

Since this code base is not very strict on type checking, it doesn't show up as linting issue but IMO it should still be fixed.

rchl replied:
Looked more into this file and nothing is really annotated with types so this issue probably doesn't matter. It should probably be a separate task to get types right in this server. So feel free to resolve this thread (I can't).

rchl commented:
I can't mark it because it's not touched in this PR but this code in `workspace.py`:

```py
        if message:
            value["message"] = message
        if percentage:
            value["percentage"] = percentage
```

should if-check against `None` since an empty message and a percentage value of 0 should be allowed.

rchl commented on pylsp/workspace.py:
```diff
@@ -152,6 +167,9 @@ def _progress_begin(
     ) -> str:
         token = str(uuid.uuid4())
+
+        self._endpoint.request(self.M_INITIALIZE_PROGRESS, {'token': token}).result()
```
This code should ideally also catch error case and disable reporting then. I guess right now if server responds with error then the corresponding functionality that triggered the reporting will not work?

syphar replied:
That would mean an editor reports the progress capability, but doesn't react to this? 

I lean towards seeing how the real world reacts to this, and fix it if it happens. 

( I'm also happy to add it here in this PR if needed)

syphar committed:
handle some RPC calls asynchronously to support progress

syphar commented:
short update here: 
when testing I discovered that after the first reported progresses the token-initialization starts to timeout, and other LSP requests too. 

I'm not sure (yet) if this is because of my editor or something else in my change. 

edit: to be more exact: progress reporting for linting works, also for new files opened. But when I try to do _go to definition_, initializing tokens times out.

syphar committed:
support empty progress messages and zero percent progress

syphar commented:
@ccordoba12 I found the issue that was blocking this. 

The async response from the create-token request wasn't received until _after_ the synchronous `definitions` request was finished. 

The solution I found was that we return a callable from `definitions`, which makes the `Endpoint` handle it in a thread, enabling us to to the parallel request handling. 

What do you think about this solution? 

Should it look differently? Should we make _all_ hook calls threaded so eventual plugin devs don't hit the same issue?

rchl commented:
I don't know pylsp code base enough to say for sure but this change to make some requests async sounds quite dangerous to be honest. I can theorize a bit and imagine that this could result in results returned from those requests getting out of sync with the document state.

For example, in theory, the references request would come but not get handled immediately and then didChange notification would come that would modify the document state. Now the processing of the `references` request might result in unexpected result because the request was made for position that no longer is what it was before.

Of course this is just me speculating since I don't know how pylsp works exactly but I imagine it might be safer to make all requests/notifications async to get more consistent and correct results. But that would still be risky, potentially for other reasons.

syphar commented:
short note before I leave for the day: the `lint` requests already were running asynchronously before this PR. This was also the reason why the token-initialize worked for linting, while it didn't for the other requests

rchl commented:
Though `lint` would be immune to the case I've mentioned because it doesn't use any input state. It just gets delayed after document was opened or modified and then works on whatever state the document is currently in when it has a chance to run.

This hypothetical issue would rather apply to requests like `references` or `definition` whose result is related to the input params provided by the client.

ccordoba12 commented:
> What do you think about this solution?

@syphar, I'm really not sure what to do about this. As @rchl said, this could generate unexpected results if the document state changed with respect to the state used by a request being processed on a thread.

Is there another way to handle this that you can think of?

syphar commented:
> > What do you think about this solution?
> 
> @syphar, I'm really not sure what to do about this. As @rchl said, this could generate unexpected results if the document state changed with respect to the state used by a request being processed on a thread.
> 
> Is there another way to handle this that you can think of?

I believe I don't know enough about JSON-RPC & LSP in general, and `pylsp_jsonrpc` specifically. 

From what I saw, I wasn't able to receive the answer for the initialize-token request, while _synchronously_ handling the request from the editor. 

So while I definitely don't rule out the possibility that there is a better way of doing this, I can't see it right now. 

I can leave this on my list, but finding a way will first need some reading & testing on how the protocol & the implementation works. If no one can provide a hint ;)

ccordoba12 commented:
> For example, in theory, the references request would come but not get handled immediately and then didChange notification would come that would modify the document state. I can theorize a bit and imagine that this could result in results returned from those requests getting out of sync with the document state.

I've run this scenario using definitions instead of references and adding `time.sleep(5)` in `pylsp_definitions` to simulate a long running operation. After requesting for an object's definition, I removed the line that contained that object to check how pylsp handled the situation.

What I found is precisely what @rchl predicted: an error in the server because the line is not available anymore.

> So while I definitely don't rule out the possibility that there is a better way of doing this, I can't see it right now.
> I can leave this on my list, but finding a way will first need some reading & testing on how the protocol & the implementation works. If no one can provide a hint ;)

@syphar, I think for this to work the server needs to be fully async-aware. There's some work to make `python-lsp-jsonrpc` async [here](https://github.com/python-lsp/python-lsp-jsonrpc/pull/12). We would need to do something similar for the server too.

And given what @rchl reported on issue #325, I think the best we can do for now is to remove progress reporting (with the exception of linting, perhaps) because it's not working as expected.

syphar commented:
> > For example, in theory, the references request would come but not get handled immediately and then didChange notification would come that would modify the document state. I can theorize a bit and imagine that this could result in results returned from those requests getting out of sync with the document state.
> 
> 
> 
> I've run this scenario using definitions instead of references and adding `time.sleep(5)` in `pylsp_definitions` to simulate a long running operation. After requesting for an object's definition, I removed the line that contained that object to check how pylsp handled the situation.
> 
> 
> 
> What I found is precisely what @rchl predicted: an error in the server because the line is not available anymore.
> 

Thank you for testing!

> So while I definitely don't rule out the possibility that there is a better way of doing this, I can't see it right now.
> I can leave this on my list, but finding a way will first need some reading & testing on how the protocol & the implementation works. If no one can provide a hint ;)
> @syphar, I think for this to work the server needs to be fully async-aware. There's some work to make python-lsp-jsonrpc async here. We would need to do something similar for the server too.

Thanks for the info, this sounds like some more work. 

> And given what @rchl reported on issue #325, I think the best we can do for now is to remove progress reporting (with the exception of linting, perhaps) because it's not working as expected.

That's up to you to decide of course. In my mind, even though we're not following the standard, progress reporting works (without this PR), with the exception of sublime text.

ccordoba12 commented:
> In my mind, even though we're not following the standard, progress reporting works (without this PR)

But the standard is there for a reason. So, to clarify my previous point: I think we should leave progress reporting, along with the changes you've done in this PR to make it LSP compliant. But we should remove it from definitions, references, formatting and renames until the server is async-aware (that would preserve linting, which is the request you need it for, right?)

syphar commented:
> > In my mind, even though we're not following the standard, progress reporting works (without this PR)
> 
> But the standard is there for a reason. So, to clarify my previous point: I think we should leave progress reporting, along with the changes you've done in this PR to make it LSP compliant. 

I assume you mean the changes from this PR, without the threading / async change. I can probably invest some time into this the next days. 

The only concern I would personally have with this is that the implementation of progress reporting would just _hang_ when used in the wrong way (in a sync handler), which could be confusing for plugin developers. I'll check if I can make it fail in these cases. 

> But we should remove it from definitions, references, formatting and renames until the server is async-aware (that would preserve linting, which is the request you need it for, right?)

Personally I needed it primarily for linting, but also for definitions. But **I totally understand your reasoning from a project viewpoint**, so that's fine with me. 

To elaborate on my need with definitions: 
in a mid-sized project I'm working on (400k LOC, + 300 dependencies), "go to definition" sometimes takes 5-10 seconds, or even longer. So after pressing the corresponding keymap, I can't quickly differentiate between the LSP not being able to find a definition, and it just taking a longer time. 

But I'll find another way, perhaps adding status reporting as a lua-wrapper, and/or showing an error message when no definition can be found.

ccordoba12 commented:
> I assume you mean the changes from this PR, without the threading / async change.

Right.

> I can probably invest some time into this the next days.

Great! Thanks for your help with that.

> The only concern I would personally have with this is that the implementation of progress reporting would just hang when used in the wrong way (in a sync handler), which could be confusing for plugin developers. I'll check if I can make it fail in these cases.

Good point. If you don't manage to make it fail, then we can add a mention to our docs (which are under the `docs` folder) to say that progress reporting doesn't work for sync handlers.

> But I'll find another way, perhaps adding status reporting as a lua-wrapper, and/or showing an error message when no definition can be found.

Ok, thanks for understanding.

syphar committed:
revert some commands back to synchronous calls

syphar committed:
re-add async formatting, it doesn't use position / ranges

syphar committed:
fix pylint warnigs

syphar committed:
remove progress reporting from unsupported plugins

syphar committed:
fix pylint errors

syphar commented:
@ccordoba12 so, I invested some time into this PR today. 

* I reverted `definitions`, `format_range` and `references` back to synchronous calls, since they get some document position 
* I left `format_document` async, since similar to `lint`, it only gets the full document. 
* I removed progress reporting for the plugins that are handled synchronously 
* ( I'm not sure about this one): I added `timeout=1.0` to the create-token request. This would kick in for the case that someone tries to do progress reporting in a synchronous handler. I didn't see any other way to get an error for this situation. This would also error when the editor is just too slow. 
* I also changed the logic, so when token-creation is failing, or timing out, we just don't do progress reporting. 

What do you think about these changes now?

ccordoba12 commented on pylsp/workspace.py:
```diff
@@ -135,15 +136,43 @@ def report_progress(
+                token = self._progress_begin(title, message, percentage)
+            except Exception:  # pylint: disable=broad-exception-caught
+                log.warning(
+                    "error while trying to initialize progress reporting."
```
```suggestion
                    "There was an error while trying to initialize progress reporting."
```

ccordoba12 commented on pylsp/workspace.py:
```diff
@@ -135,15 +136,43 @@ def report_progress(
-        finally:
-            self._progress_end(token)
+        # FALLBACK:
+        # if the client doesn't support progress reporting, or if we failed to
```
```suggestion
        # If the client doesn't support progress reporting, or if we failed to
```

ccordoba12 commented on pylsp/python_lsp.py:
```diff
@@ -358,7 +360,7 @@ def execute_command(self, command, arguments):
 
     def format_document(self, doc_uri, options):
-        return self._hook('pylsp_format_document', doc_uri, options=options)
+        return lambda: self._hook('pylsp_format_document', doc_uri, options=options)
```
What are the advantages of making this call async?

syphar replied:
it's possible to keep progress reporting for it.

ccordoba12 replied:
Ok, I understand now: you removed it for `format_range` but leave it for `format_document`. That's fine by me.

ccordoba12 reviewed (commented):
A couple of small suggestions and a question, otherwise looks good to me @syphar.

ccordoba12 commented:
> What do you think about these changes now?

I think this is almost ready. Thanks @syphar for your continued work on this!

@rchl, could you test this PR against Sublime to see if things work as expected now? Thanks!

rchl commented:
If the client provides a `workDoneToken` in the request then it can be used instead of creating new token.

I guess it's technically not an error to still create a token like this code does but just saying.

rchl commented on pylsp/workspace.py:
```diff
@@ -135,15 +136,43 @@ def report_progress(
+            except Exception:  # pylint: disable=broad-exception-caught
+                log.warning(
+                    "error while trying to initialize progress reporting."
+                    "Either the editor failed, was too slow to answer, or this LSP "
```
There should never be a need for a timeout. Client is supposed to respond with an Error if it doesn't understand the request. But since you are checking the capability then that should not happen even.

EDIT: Unless it's a workaround for a deadlock when triggering from synchronous requests? But in that case the line below should be more relevant anyway.

syphar replied:
The second reason for the timeout is plugin developers. 

Assuming a plugin dev is using the progress reporting in a synchronous handler. 

Without the timeout this request would just hang, and the response for the create-token would only be received after the handler is finished. 

Without knowing more about internals of the lsp protocol, and this library, I didn't see an other way to prevent the request just hanging.

syphar replied:
Seeing your edit, what do you propose?

rchl replied:
OK, but then the next line of this warning is more relevant anyway?

syphar replied:
ah, so it's about the error message, right? 

I updated it, I hope better to understand

rchl commented on pylsp/plugins/definition.py:
```diff
@@ -8,26 +8,25 @@
-            follow_imports=settings.get('follow_imports', True),
-            follow_builtin_imports=settings.get('follow_builtin_imports', True),
-            **code_position)
+def pylsp_definitions(config, document, position):
+    settings = config.plugin_settings('jedi_definition')
+    code_position = _utils.position_to_jedi_linecolumn(document, position)
```
Did we decide to remove progress reporting from this one? Since this one is explicitly triggered by the user and can take longer time, it feels like it should be OK to have it here IMO.

syphar replied:
Both changes come from how I understood the comment by @ccordoba12 : 
> > For example, in theory, the references request would come but not get handled immediately and then didChange notification would come that would modify the document state. I can theorize a bit and imagine that this could result in results returned from those requests getting out of sync with the document state.
> 
> I've run this scenario using definitions instead of references and adding `time.sleep(5)` in `pylsp_definitions` to simulate a long running operation. After requesting for an object's definition, I removed the line that contained that object to check how pylsp handled the situation.
> 
> What I found is precisely what @rchl predicted: an error in the server because the line is not available anymore.

and then 

> > In my mind, even though we're not following the standard, progress reporting works (without this PR)
> 
> But the standard is there for a reason. So, to clarify my previous point: I think we should leave progress reporting, along with the changes you've done in this PR to make it LSP compliant. But we should remove it from definitions, references, formatting and renames until the server is async-aware (that would preserve linting, which is the request you need it for, right?)

If that's wrong, I can change it back of course :)

ccordoba12 replied:
> Did we decide to remove progress reporting from this one? Since this one is explicitly triggered by the user and can take longer time, it feels like it should be OK to have it here IMO.

Yes, we did. See [my comment](https://github.com/python-lsp/python-lsp-server/pull/328#issuecomment-1474940689) above for the reasons (basically, it gives an error in the server, as you predicted).

ccordoba12 replied:
> If that's wrong, I can change it back of course :)

Please don't. It's fine for now.

rchl replied:
Fine by me. BTW. I can't resolve issues created by myself so feel free to do that for me.

rchl commented on pylsp/plugins/references.py:
```diff
@@ -8,20 +8,19 @@
-    with workspace.report_progress("references"):
-        code_position = _utils.position_to_jedi_linecolumn(document, position)
-        usages = document.jedi_script().get_references(**code_position)
+def pylsp_references(document, position, exclude_declaration=False):
+    code_position = _utils.position_to_jedi_linecolumn(document, position)
+    usages = document.jedi_script().get_references(**code_position)
```
Also here? This one is explicitly triggered by the user.

syphar replied:
see above

rchl commented:
BTW. We've implemented a workaround in ST some time ago for servers that don't create a token so it was not a problem anymore. But those changes appear to work fine too.

syphar committed:
fix typos

syphar committed:
update warning message when progress coudln't be initialized

ccordoba12 commented:
@rchl, are you ok with this now? Just checking before merging.

rchl commented:
yeah, fine by me

ccordoba12 reviewed (approved):
Thanks @syphar!
